### PR TITLE
8260267: ZGC: Reduce mark stack usage

### PR DESCRIPTION
**Pull request by pliden**
Under some conditions, striped marking can cause excessive mark stack memory usage. For example, when the object graph has one or more N:1 relations ships, where N is a large number.

To avoid pushing mark stack entries to already marked objects, and thus reduce mark stack memory usage, this patch adds does two things:
 * GC threads will under some conditions mark the object before pushing, else (if the condition to mark failed)
 * GC threads will check if the object is already marked before pushing.

The conditions for marking before pushing is:
 * We are using a single stripe, in which cases striped marking is essentially disabled anyway.
 * We see excessive mark stack usage. In other words, the mark stack allocator indicates "high usage".

To make this easier to review, this PR contains two commits:
 * `ZGC: Reduce mark stack usage` - This patch adds the infrastructure needed to mark before push, but only does so if we are using a single stripe.
 * `ZGC: Mark before push if mark stack usage is high` - This patch adds the condition to mark before push if the mark stack allocator indicates "high usage". As a bonus, this patch will also uncommit memory used for mark stacks after the marking has completed.

Testing:
 * Tier1-7
 * SPECjbb2015, score and marking times unaffected
 * casparcwang@tencent.com reports that this patch, in combination with PR #3455, solves the problem for Apache Zookeeper.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260267](https://bugs.openjdk.java.net/browse/JDK-8260267): ZGC: Reduce mark stack usage

### Reviewers
 * [Stefan Johansson](https://openjdk.java.net/census#sjohanss) (@kstefanj - **Reviewer**) ⚠️ Review applies to 2266d9d1c1b32d0a0d4fedae32118423f2bc0c5d
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - Committer) ⚠️ Review applies to 77a65449a0a4a7cf4eee177036a114592f7eb088

### Contributors
 * Wang Chao `<wchao@openjdk.org>`
 * Hui Shi `<hshi@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3460/head:pull/3460` \
`$ git checkout pull/3460`

Update a local copy of the PR: \
`$ git checkout pull/3460` \
`$ git pull https://git.openjdk.java.net/jdk pull/3460/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3460`

View PR using the GUI difftool: \
`$ git pr show -t 3460`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3460.diff">https://git.openjdk.java.net/jdk/pull/3460.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/z/zMarkStackAllocator.cpp [hunk 1]
```diff
@@ -55,12 +55,74 @@ ZMarkStackSpace::ZMarkStackSpace() :
...
 bool ZMarkStackSpace::is_initialized() const {
   return _start != 0;
 }
 
+size_t ZMarkStackSpace::size() const {
+  return _end - _start;
+}
+
+static bool is_high_usage(size_t size) {
+  // Consider usage to be high if we've used more than a 8th of the available space.
```
Comment (kstefanj) [inline src/hotspot/share/gc/z/zMarkStackAllocator.cpp:72]: "a 8th" -> "an 8th" or "one 8th"

File: src/hotspot/share/gc/z/zMarkStackAllocator.cpp [hunk 2]
```diff
@@ -55,12 +55,74 @@ ZMarkStackSpace::ZMarkStackSpace() :
...
+          "maximum number of bytes allocated for mark stacks. Current limit is " SIZE_FORMAT "M.",
+          ZMarkStackSpaceLimit / M);
+  }
+
+  if (!_high_usage && is_high_usage(new_size)) {
+    Atomic::store(&_high_usage, true);
+  }
+
+  log_debug(gc, marking)("Expanding mark stack space: " SIZE_FORMAT "M->" SIZE_FORMAT "M (%s Usage)",
+                         old_size / M, new_size / M, _high_usage ? "High" : "Low");
```
Comment (kstefanj) [inline src/hotspot/share/gc/z/zMarkStackAllocator.cpp:101]: A bit strange to only have "Low" and "High" usage, maybe "Normal" and "High" would be more accurate.

**Review by kstefanj — APPROVED** [review]
Looks good, just two minor comments.

**Review by albertnetymk — APPROVED** [review]

File: src/hotspot/share/gc/z/zMarkStackAllocator.cpp (hunk 1 above)
Comment (pliden) [reply src/hotspot/share/gc/z/zMarkStackAllocator.cpp]: Fixed!

File: src/hotspot/share/gc/z/zMarkStackAllocator.cpp (hunk 2 above)
Comment (pliden) [reply src/hotspot/share/gc/z/zMarkStackAllocator.cpp]: Fixed!

**Comment by huishi-hs** [comment]
@pliden Is this PR still under testing or needs more reivew?  Hope this doesn't distrub you from what you're currently working on.

We would like to backport this PR after it get integrated.

**Comment by pliden** [comment]
It's still under testing and discussion. I plan to push an update branch to this later today.

**Comment by pliden** [comment]
Updated PR with the following:
1) Always mark before push. After additional testing and some discussions, it turns out that the overhead of marking before pushing it's that high. Concurrent Mark times increase by about 1% or so on SPECjbb2015 and BigRamTester. So let's keep thing simple, and avoid introducing that arbitrary "high usage" threshold.
2) Only uncommit the part of the mark stack space that hasn't been used.
3) Print mark stack memory usage in the GC log.

**Review by kstefanj — APPROVED** [review]
Still good or rather even better. Nice to avoid having the high usage state.

**Comment by huishi-hs** [comment]
> Updated PR with the following:
> 
> 1. Always mark before push. After additional testing and some discussions, it turns out that the overhead of marking before pushing isn't that high. Concurrent Mark times increase by about 1% or so on SPECjbb2015 and BigRamTester. So let's keep thing simple, and avoid introducing that arbitrary "high usage" threshold.
> 2. Only uncommit the part of the mark stack space that hasn't been used.
> 3. Print mark stack memory usage in the GC log.

This is great! entire code is unified with always before push. Thanks~!

**Comment by pliden** [comment]
Thanks @albertnetymk and @kstefanj for reviewing!

/integrate

**Pull request closed without merging**